File: application/app/domain/model/filters/BlacklistIps.scala
package domain.model.filters

import common.validation.constraints.TraversableSize
import domain.model.{Filter, SetValue}
import javax.validation.constraints.NotNull
import play.mvc.Http.HeaderNames.X_FORWARDED_FOR

import scala.annotation.meta.field

case class BlacklistIps(
                         @(NotNull @field) @(TraversableSize @field)(max = 255) blacklist: Set[String]
                       ) extends Filter {
  override val name: String = Filter.BLACKLIST_IPS
  override val value: SetValue = SetValue(blacklist)
  override val filter: Map[String, String] => Boolean = {
    headers => !headers.contains(X_FORWARDED_FOR) || !blacklist.contains(headers(X_FORWARDED_FOR))
  }
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/model/filters/BlacklistIps.java
package com.cosmin.wsgateway.domain.model.filters;

import com.cosmin.wsgateway.domain.model.Filter;

import lombok.Value;

import java.util.Map;
import java.util.Set;

@Value
public class BlacklistIps implements Filter<Set<String>> {
    private final Set<String> hosts;

    @Override
    public String name() {
        return "blacklistIps";
    }

    @Override
    public Set<String> value() {
        return hosts;
    }

    @Override
    public boolean isAllowed(Map<String, String> requestHeaders) {
        return !requestHeaders.containsKey("Host") || !hosts.contains(requestHeaders.get("Host"));
    }
}
